### PR TITLE
DAOS-14155 gurt: Add new logging macros and tools to support them.

### PR DESCRIPTION
**Pull request by ashleypittman**
Required-githooks: true

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Review by jolivier23 — APPROVED** [review]

File: src/include/gurt/debug.h [hunk 1]
```diff
@@ -197,6 +195,32 @@ extern void (*d_alt_assert)(const int, const char*, const char*, const int);
...
+	} while (0)
+
+#define DL_ERROR(_rc, _fmt, ...)                                                                   \
+	do {                                                                                       \
+		D_DEBUG(DLOG_ERR, _fmt ": " DF_RC "\n", ##__VA_ARGS__, DP_RC(_rc));                \
+	} while (0)
+
+#define DHL_WARN(_desc, _rc, _fmt, ...)                                                            \
+	do {                                                                                       \
+		_D_DEBUG(_D_TRACE_NOCHECK, DLOG_WARN, (_desc), _fmt ": " DF_RC "\n",               \
```
Comment (jolivier23) [inline src/include/gurt/debug.h:215]: What is the intention of DHL * ?

Comment (ashleypittman) [reply src/include/gurt/debug.h:215]: I can't remember how the names came about but it adds a descriptor so you can trace error/debug messages to where they originate from.  The idea is once you get a DL_ERROR call for a descriptor then you can trace back and highlight all the debug messages from it.

Two problems with this though, it's not possible to automatically find the right pointer to use as `desc` so code will need updating manually and the above only works if debug messages are updated as well as error messages.

**Review by kjacque — APPROVED** [review]

**Review by jolivier23 — APPROVED** [review]

**Comment by brianjmurrell** [comment]
The last 3 builds of this job have created a massive amount of data (sorted by size):

```
7.5M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/4
15M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/3
417M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/1
518M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/5
755M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/6
848M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/7
2.1G	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/2
56G	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/10
56G	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/8
56G	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9
```

Looking at the biggest one of these, build number 9:

```
201M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9/archive/nlt_logs/el8.fault-injection/dnt_fi_pool-list-containers_logs
211M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9/archive/nlt_logs/el8.fault-injection/dnt_fi_cont-list-attr_logs
217M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9/archive/nlt_logs/el8.fault-injection/dnt_fi_cont-get-attr_logs
227M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9/archive/nlt_logs/el8.fault-injection/dnt_fi_cont-get-prop_logs
237M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9/archive/nlt_logs/el8.fault-injection/dnt_fi_cont-create_logs
242M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9/archive/nlt_logs/el8.fault-injection/dnt_fi_cont-query_logs
393M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9/archive/nlt_logs/el8.fault-injection/dnt_fi_filesystem-copy-trunc_logs
575M	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9/archive/nlt_logs/el8.fault-injection/dnt_fi_dfuse_logs
2.0G	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9/archive/nlt_logs/el8.fault-injection/dnt_fi_filesystem-copy_logs
48G	jobs/daos-stack/jobs/daos/branches/PR-12800/builds/9/archive/nlt_logs/el8.fault-injection/dnt_fi_cont-check_logs
```

And digging into the `el8.fault-injection/dnt_fi_cont-check_logs` there is no one great big log but there are 8638 logs ranging from 460K to 12M.

Is this new amount of logging expected to become the norm when this lands?  We will have to do some planning and restructuring if this is expected to be the new norm as this is massively bigger than the amount of logs we are currently storing for our jobs.

**Comment by ashleypittman** [comment]
> Is this new amount of logging expected to become the norm when this lands? We will have to do some planning and restructuring if this is expected to be the new norm as this is massively bigger than the amount of logs we are currently storing for our jobs.

No, those size changes are no the result of this PR, they were a different problem which should now be resolved.

**Pull request merged**